### ISSUE TITLE
[Android] Remove selected components from globeView

### ISSUE THREAD

**Issue by cristianabraga** [mousebird/WhirlyGlobe#1162]
Hi guys! I have a globe with the following test case:

When I press on country A, I colored that country, when I press on country B, get colored also.
 
Then I want when I double press to Country A to removed filled color and also from selectedCompornents. My problem is that using globeControl.removeObject(selectedComponentObject, MaplyBaseController.ThreadMode.ThreadAny); I always removing country B and not A.

Could you please help me how to specify which vectorObject to remove? An example will be great.

Thanks a lot.

**Comment by mousebird** [comment]
You pass in the ComponentObject that was returned for country A.  That should remove just that country.

If you added both countries at once, you need to add them individually and track each ComponentObject.

**Comment by cristianabraga** [comment]
Thanks for your answer!! I have tried to implement that you described to me but still not successfully :( 
Could you please take a look to the class, I'd be very grateful

`public class HelloGlobeFragment extends GlobeMapFragment {

    ComponentObject selectedComponentObject;
    CheckBox been;
	...

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle inState) {
        super.onCreateView(inflater, container, inState);

        return baseControl.getContentView();
    }

    @Override
    protected void controlHasStarted() {       
		....
		globeControl.gestureDelegate = this;

        VectorsJson vectorsJson = new VectorsJson();
        try {
            vectorsJson.overlayCountries(globeControl, getContext());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void userDidSelect(GlobeController mapControl, SelectedObject[] selObjs, Point2d loc, Point2d screenLoc) {
        String msg = "Selected feature count: " + selObjs.length;
        for (SelectedObject obj : selObjs) {
            // GeoJSON
            if (obj.selObj instanceof VectorObject) {

                final VectorObject vectorObject = (VectorObject) obj.selObj;
                AttrDictionary attributes = vectorObject.getAttributes();
                final String adminName = attributes.getString("ADMIN");
                msg += "\nCountry: " + adminName;

                showPopup(adminName);

                been.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
				
                    @Override
                    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                        if (buttonView.isChecked()) {
						
                            drawVectorObjectAsSelected(vectorObject,Color.GREEN);
                            hashMap.put(adminName,Color.GREEN);
                            MainActivity.myDialog.dismiss();
							
                        }else if(!buttonView.isChecked()){
						
                            globeControl.removeObject(selectedComponentObject,MaplyBaseController.ThreadMode.ThreadAny);
							
                            been.setChecked(false);

                            hashMap.remove(adminName);
                            MainActivity.myDialog.dismiss();
                        }
                    }
                });
            }
        }
        Toast.makeText(getActivity(), msg, Toast.LENGTH_LONG).show();
        }

    public void showPopup(String country) {
			...

    }

    public void drawVectorObjectAsSelected(VectorObject vectorObject,Integer color) {
                
        VectorInfo vectorInfo = new VectorInfo();
        vectorInfo.setColor(color);
        vectorInfo.setFilled(true);
        vectorInfo.setLineWidth(7.f);
        vectorInfo.setDrawPriority(Integer.MAX_VALUE);
		
        globeControl.addVector(vectorObject,vectorInfo, MaplyBaseController.ThreadMode.ThreadAny);
   }

}`

**Comment by mousebird** [comment]
I don't see where you're setting selectedComponentObject.  Make sure that's set to the return of addVector

**Comment by cristianabraga** [comment]
Ok, got it, thank you. I've added (selectedComponentObject = globeControl.addVector(vectorObject,vectorInfo......)

Should I work with a list of ComponentObject? Because the issue is still persisting: Click on Country A, Click on Country B, then Click again on country A, and it is disappearing B.

**Comment by mousebird** [comment]
I'd imagine you'd need a list or array of them if you want to track each country individually.  Without digging into your source it's hard to say.

I can tell you that addVector returns a handle (ComponentObject) for what you added with addVector.  Removing that will remove whatever you created in that call.

**Comment by cristianabraga** [comment]
Me again, with the full source code this time. Basically I have only 2 classes: VectorsJson (where I am displaying all countries): 

`public class VectorsJson {

    private ArrayList<VectorObject> vectors = new ArrayList<>();

    public VectorsJson() {

    }

    public void overlayCountries(MaplyBaseController controller, Context context) throws Exception{
        VectorInfo vectorInfo = new VectorInfo();
        vectorInfo.setColor(Color.RED);
        vectorInfo.setLineWidth(4.f);

        AssetManager assetMgr = context.getAssets();
        String[] paths = assetMgr.list("country_json_50m");

        for (String path:paths){
            System.out.print(path);
            InputStream stream = assetMgr.open("country_json_50m/" + path);

            try{
                VectorObject vecObject = new VectorObject();
                vecObject.selectable = true;
                String json = IOUtils.toString(stream, Charset.defaultCharset());
                if (vecObject.fromGeoJSON(json)) {
                    vectors.add(vecObject);
                }
            }
            finally {
                try {
                    stream.close();
                } catch (IOException e) {
                }

            }
        }
        ComponentObject compObj = controller.addVectors(vectors, vectorInfo, MaplyBaseController.ThreadMode.ThreadAny);
        VectorInfo newVectorInfo = new VectorInfo();
        newVectorInfo.setColor(Color.WHITE);
        newVectorInfo.setLineWidth(5.f);
        controller.changeVectors(compObj,newVectorInfo, MaplyBaseController.ThreadMode.ThreadAny);
    }

}`


And the other one "HelloGlobeFragment"

`
public class HelloGlobeFragment extends GlobeMapFragment {
    ComponentObject selectedComponentObject;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle inState) {
        super.onCreateView(inflater, container, inState);

        return baseControl.getContentView();
    }

    @Override
    protected MapDisplayType chooseDisplayType() {
        return MapDisplayType.Globe;
    }

    @Override
    protected void controlHasStarted() {

        String cacheDirName = "stamen_watercolor";
        File cacheDir = new File(getActivity().getCacheDir(), cacheDirName);
        cacheDir.mkdir();
        RemoteTileSource remoteTileSource = new RemoteTileSource(new RemoteTileInfo("http://tile.stamen.com/watercolor/", "png", 0, 18));
        remoteTileSource.setCacheDir(cacheDir);
        SphericalMercatorCoordSystem coordSystem = new SphericalMercatorCoordSystem();

        QuadImageTileLayer baseLayer = new QuadImageTileLayer(globeControl, coordSystem, remoteTileSource);
        baseLayer.setImageDepth(1);
        baseLayer.setSingleLevelLoading(false);
        baseLayer.setUseTargetZoomLevel(true);
        baseLayer.setCoverPoles(true);
        baseLayer.setHandleEdges(true);

        globeControl.addLayer(baseLayer);
        globeControl.gestureDelegate = this;

        VectorsJson vectorsJson = new VectorsJson();
        try {
            vectorsJson.overlayCountries(globeControl, getContext());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void userDidSelect(GlobeController mapControl, final SelectedObject[] selObjs, Point2d loc, Point2d screenLoc) {
        String msg = "Selected feature count: " + selObjs.length;
        for (SelectedObject obj : selObjs) {
            if (obj.selObj instanceof VectorObject) {

                final VectorObject vectorObject = (VectorObject) obj.selObj;
                AttrDictionary attributes = vectorObject.getAttributes();
                final String adminName = attributes.getString("ADMIN");
                msg += "\nCountry: " + adminName;

                drawVectorObjectAsSelected(vectorObject,Color.RED);


            }
        }
        Toast.makeText(getActivity(), msg, Toast.LENGTH_LONG).show();
        }


    @Override
    public void userDidLongPress(GlobeController globeController, SelectedObject[] selObjs, Point2d loc, Point2d screenLoc) {
        globeControl.removeObject(selectedComponentObject, MaplyBaseController.ThreadMode.ThreadAny);
        Toast.makeText(getActivity(), "Trying to remove a specific country and not to have anymore a RED color...", Toast.LENGTH_LONG).show();
    }


    public void drawVectorObjectAsSelected(VectorObject vectorObject,Integer color) {

        VectorInfo vectorInfo = new VectorInfo();
        vectorInfo.setColor(color);
        vectorInfo.setFilled(true);
        vectorInfo.setLineWidth(7.f);
        vectorInfo.setDrawPriority(Integer.MAX_VALUE);
        selectedComponentObject = globeControl.addVector(vectorObject,vectorInfo, MaplyBaseController.ThreadMode.ThreadAny);
   }

}
`

In a nutshell: I want at first click to color in RED (ex: Spain) and for long press I want remove this country (not to be red anymore). My issue is persisting when I color many countries and then at long press I want to disable only a specific one (ex: Spain). 

Please, need an advice what should I change into the code :(. 
I will remain grateful forever to you.
